kotlinx-coroutines-core/common/test/CoroutineScopeTest.kt

```diff
@@ -247,6 +247,20 @@ class CoroutineScopeTest : TestBase() {
         finish(7)
     }
 
+    @Test
+    fun testNewCoroutineContextDispatcher() {
+        fun newContextDispatcher(c1: CoroutineContext, c2: CoroutineContext) =
+            ContextScope(c1).newCoroutineContext(c2)[ContinuationInterceptor]
+
+        assertSame(Dispatchers.Default, newContextDispatcher(EmptyCoroutineContext, EmptyCoroutineContext))
+        assertSame(Dispatchers.Default, newContextDispatcher(EmptyCoroutineContext, Dispatchers.Default))
+        assertSame(Dispatchers.Default, newContextDispatcher(Dispatchers.Default, EmptyCoroutineContext))
+        assertSame(Dispatchers.Default, newContextDispatcher(Dispatchers.Default, Dispatchers.Default))
+        assertSame(Dispatchers.Default, newContextDispatcher(Dispatchers.Unconfined, Dispatchers.Default))
+        assertSame(Dispatchers.Unconfined, newContextDispatcher(Dispatchers.Default, Dispatchers.Unconfined))
+        assertSame(Dispatchers.Unconfined, newContextDispatcher(Dispatchers.Unconfined, Dispatchers.Unconfined))
+    }
```

Comment:
I think we also need to check the user-visible bug directly, like what you have in your reproducer:
```kotlin
    runTest {
        val job = Job()
        val scope = CoroutineScope(job + coroutineContext[CoroutineExceptionHandler]!!)
        scope.launch(Dispatchers.Default) {
            assertSame(Dispatchers.Default, coroutineContext[ContinuationInterceptor])
        }
        scope.launch {
            assertSame(Dispatchers.Default, coroutineContext[ContinuationInterceptor])
        }
        job.complete()
        job.join()
    }
```